src/dhcpmon/src/main.c

```diff
@@ -0,0 +1,205 @@
+/**
+ * @file main.c
+ *
+ *  @brief: Main entry point for dhcpmon utility.
+ *
+ */
+
+#include <err.h>
+#include <errno.h>
+#include <libgen.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <string.h>
+#include <semaphore.h>
+#include <syslog.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "dhcp_mon.h"
+#include "dhcp_devman.h"
+
+/** dhcpmon_active: semaphore main thread wait on before exiting */
+static sem_t dhcpmon_active;
+/** dhcpmon_default_snaplen: default snap length of packet being captured */
+static const uint32_t dhcpmon_default_snaplen = 65535;
+/** dhcpmon_default_pcap_timout_ms: default timeout of libpcap utility before delivering packets to awaiting thread */
+static const uint32_t dhcpmon_default_pcap_timout_ms = 100;
+/** dhcpmon_default_health_check_window: default value for a time window, during which DHCP DORA packet counts are being collected */
+static const uint32_t dhcpmon_default_health_check_window = 12;
+/** dhcpmon_default_unhealthy_max_count: default max consecutive unhealthy status reported before reporting an issue with DHCP relay */
+static const uint32_t dhcpmon_default_unhealthy_max_count = 10;
+
+/**
+ * @code usage(prog);
+ *
+ * @brief prints help message about how to use dhcpmon utility
+ *
+ * @param prog program name
+ *
+ * @return none
+ */
+static void usage(const char *prog)
+{
+    printf("Usage: %s -id <south interface> {-iu <north interface>}+ [-w <snapshot window in sec>]"
+            "[-c <unhealthy status count] [-s <snap length>] [-t <pcap timeout>] [-d]\n", prog);
+    printf("where\n");
+    printf("\tsouth interface: is a vlan interface,\n");
+    printf("\tnorth interface: is a TOR-T1 interface,\n");
+    printf("\tsnapshot window: during which DHCP counters are gathered and DHCP status is validated (default %d),\n",
+            dhcpmon_default_health_check_window);
+    printf("\tunhealthy status count: count of consecutive unhealthy status before writing an alert to syslog "
+           "(default %d),\n",
+           dhcpmon_default_unhealthy_max_count);
+    printf("\tsnap length: snap length of packet capture (default %d),\n", dhcpmon_default_snaplen);
+    printf("\tpcap timeout: pcap library time out (default %d).\n", dhcpmon_default_pcap_timout_ms);
+    printf("\t-d: daemonize %s.\n", prog);
+    exit(0);
+}
+
+/**
+ * @code dhcpmon_signal_handler(signal);
+ *
+ * @brief signal handler for dhcpmon. It will initiate shutdown when signal is caught
+ *
+ * @param signal signal code that we received
+ *
+ * @return none
+ */
+static void dhcpmon_signal_handler(int signal)
+{
+    dhcp_devman_print_status();
+    dhcp_devman_stop_capture();
+    dhcp_mon_shutdown();
+    sem_post(&dhcpmon_active);
+}
+
+/**
+ * @code dhcpmon_daemonize();
+ *
+ * @brief make this utility run as a daemon.
+ *
+ * @return none
+ */
+static void dhcpmon_daemonize()
+{
```

Comment:
suggest to use libdaemon library for that also as for signals

Reply:
Did not know it existed. Will look into it. Also, after testing on target, the daemonization code is not required

Reply:
Correct. As discussed, supervisord spawns and manages all processes in our containers, and requires them to run in the foreground in order to properly watch them. Therefore, daemonization is unnecessary.

However, since it is already added, do not remove it; consider using libdaemon, but this is not a priority.

Reply:
Libdaemon does not seem to strong adoption. It has some sort of handshaking between parent of child process that adds more code. I will opt to have in this simple daemonization code.

Reply:
I'm fine with the existing daemonization code; it is pretty much in line with many existing implementations. Plus, as mentioned earlier, it most likely will not get used in SONiC.